### PR TITLE
bug/SWTCH-1503 duplicate did

### PR DESCRIPTION
**Pull request by JGiter**
- [x] test updating of did document
 - all mocked components of `DidService` with exception of `DidStore`, which is currenly undergoing changes, are replaced by exact types
 - removed several tests which are covered by other tests or not relevant to API of `DidService`

### CODE REVIEW

**Comment by jrhender** [comment]
@JGiter  `.test.env` should be named `.env.test` to align with the other `.env` files in the repo

Could you also explain how `.test.env` is intended to work with the tests please? Do the tests automatically use `.test.env` or does it need to be copied to be the `.env` file?

**Review by jrhender — COMMENTED** [review]
Overall the changes are looking pretty good.

A general suggestion I have is that it might make sense to split this PR into two different PRs. In general, I think one should not refactor (including tests) and change functionality in the same PR. This point is expressed, for example, by Martin Fowler in Workflows of Refactoring (https://martinfowler.com/articles/workflowsOfRefactoring/#2hats, https://www.youtube.com/watch?v=vqEg37e4Mkw&ab_channel=SIGSDATACOM).
In other words, the refactoring changes to `did.services.spec` should be in a separate PR from the changes to `did.service`. What are your thoughts on this?

**Comment by JGiter** [comment]
> Could you also explain how `.test.env` is intended to work with the tests please? 

I would like tests to be as close to app as possible. So I put environment variables in .env and configured test module to read them. Probably also this .env better be called `.env.spec` to make clear what it is used for

File: src/modules/did/did.service.ts [hunk 1]
```diff
@@ -338,7 +358,7 @@ export class DIDService implements OnModuleInit {
       });
       // Only refreshing a DID that is already cached.
       // Otherwise, cache could grow too large with DID Docs that aren't relevant to Switchboard
-      if (didDocEntity) {
+      if (didDocEntity !== null) {
```
[the remark this answers is not in the document]
Comment (JGiter) [reply src/modules/did/did.service.ts]: I just wanted to be more specific suspecting that this is the cause of problem I had with testing. `didDocEntity` can not be `undefined`, because `findById` returns `DIDDocumentEntity` or `null`. But test was failing for another reason, so it is not necessary to complicate this check. I'll revert it

File: src/modules/did/did.service.ts [hunk 2]
```diff
@@ -203,7 +211,13 @@ export class DIDService implements OnModuleInit {
         logs: JSON.stringify(logs),
       });
 
-      return this.didRepository.save(updatedEntity);
+      span = transaction?.startChild({
+        op: 'persisting_updated_document',
+        description: 'Resolve not cached claims',
+      });
+      const doc = await this.didRepository.save(updatedEntity);
+      span?.finish();
```
[the remark this answers is not in the document]
Comment (JGiter) [reply src/modules/did/did.service.ts]: I misunderstood purpse of transaction. I wanted to have more detailes of error originated in did service but reported from assets service. I'll revert this

**Comment by JGiter** [comment]
> To run these tests, does one also need to have the dependencies running via Docker as well?

Yes, unit tests require Docker containers. They are configured for CI tests. Locally though containers should be ran manually, but this is what we have been doing, because other unit tests, like organization, also requires containers

**Comment by jrhender** [comment]
> Yes, unit tests require Docker containers. They are configured for CI tests. Locally though containers should be ran manually, but this is what we have been doing, because other unit tests, like organization, also requires containers

Ok, thanks. So as a first step, to run the tests locally, one should start a docker compose in order to have the dependencies running, right? If so, do you know if we have that docker-compose file available in the repo?
It seems to me like we are missing the file as `docker-compose.dev.yml` & `docker-compose.prod.yml` don't seem to fit as I think we would need a `compose` file without the `app` `service`. Does this seem right?

**Comment by JGiter** [comment]
>  do you know if we have that docker-compose file available in the repo?

No, we don't have Docker config for unit tests. It would be usefull to have it along with instruction how to test in different environments

**Comment by jrhender** [comment]
Thanks @JGiter . I was able to run the tests locally.

I'm wondering if maybe it would make sense to make the tests `e2e` tests and keep most (maybe all) of the dependencies mocked in the "unit" tests.
To me, having unit tests be easy and quick to run is valuable and, I believe, that most developer would expect that one wouldn't need to run docker in order run unit tests.
These tests would need to change because in `e2e` tests, one can only call the HTTP endpoints and I don't think spies would work. But I think we can get the same value of the tests you've written (i.e. testing the DID module against an actual database) but we would keep a more intuitive "e2e" & "unit" test separation.
What are your thoughts here?

**Comment by JGiter** [comment]
> I'm wondering if maybe it would make sense to make the tests `e2e` tests and keep most (maybe all) of the dependencies mocked in the "unit" tests

I agree. This is right separation

**Comment by jrhender** [comment]
> > I'm wondering if maybe it would make sense to make the tests `e2e` tests and keep most (maybe all) of the dependencies mocked in the "unit" tests
> 
> I agree. This is right separation

Ok, sounds good. Do you think it makes to action this in this PR then? (In other words, should this PR be modified further before merging in your opinion?)

**Comment by JGiter** [comment]
Yes, I am modifing it now

**Review by jrhender — APPROVED** [review]
Great work, this is fantastic, thanks @JGiter
